R/DBI-pool.R

```diff
@@ -47,8 +48,39 @@ setMethod("onDestroy", "DBIConnection", function(object) {
 #' @export
 #' @rdname object
 setMethod("onValidate", "DBIConnection", function(object) {
-  check <- dbGetQuery(object, "SELECT 1")
-  df <- data.frame(1)
-  names(df) <- "1"
-  (check == df)[1,]
+  pool <- attr(object, "..metadata", exact = TRUE)$..pool
+  query <- pool$stateEnv$validateQuery
+
+  if (!is.null(query)) {
+    error <- try({
+      dbGetQuery(object, query)
+      return()
+    }, silent = TRUE)
+  } else {
+
+    ## options mostly gathered from here:
+    ## http://stackoverflow.com/a/3670000/6174455
+    options <- c(
+      "SELECT 1",
+      "SELECT 1 FROM DUAL",
+      "SELECT 1 FROM INFORMATION_SCHEMA.SYSTEM_USERS",
+      "SELECT * FROM INFORMATION_SCHEMA.TABLES",
+      "VALUES 1",
+      "SELECT 1 FROM SYSIBM.SYSDUMMY1",
+      "select count(*) from systables"
+    )
+
+    for (opt in options) {
+      error <- try({
+        dbGetQuery(object, opt)
+        pool$stateEnv$validateQuery <- opt
+        return()
+      }, silent = TRUE)
+    }
```

Comment:
This for-loop could use a comment--especially worth calling out the facts that 1) we return when a query option succeeds, and 2) the successful query is cached.